miurla/morphic
# feat: Implement SearchXNG as alternative search backend

# Add SearchXNG Support

## Motivation

This PR introduces support for SearchXNG as an alternative search engine in Morphic. This addition aims to provide users with more flexibility in choosing their preferred search backend and enhances privacy options by allowing self-hosted search instances.

## Summary of Changes

- Implemented a `searchXNGSearch` function in `lib/agents/tools/search.tsx` to handle SearchXNG API requests.
- Updated the search tool to support SearchXNG alongside existing Tavily and Exa search options.
- Added configuration options in `docker-compose.yml` and `.env.local.example` to set up and configure SearchXNG.
- Enhanced error handling and logging for SearchXNG API calls.
- Added new Types lib/types/index.ts for changes implemented.
- Updated README to reflect the new SearchXNG option.

## Testing Instructions

1. Clone the updated repository and install dependencies.
2. Copy `.env.local.example` to `.env.local` and set the following variables:

	- SEARCHXNG_API_URL=http://localhost:8080
	
	- SEARXNG_SECRET=your_secret_key_here 
	
	- SEARXNG_PORT=8080

	- SEARXNG_IMAGE_PROXY=true
	
	- SEARCH_API=searchxng

	- SEARXNG_BIND_ADDRESS=0.0.0.0

	- SEARXNG_LIMITER=false

3. Run `docker-compose build --no-cache && docker-compose up -d && docker-compose logs -f` 

   to start the Morphic stack with SearchXNG.

4. To test locally with node. Clone the above repo as above
- bun install to install the relevant packages
- You could run a temp searxng docker instance`docker run   -p  8080:8080 -v ./searxng-settings.yml:/etc/searxng/settings.yml searxng/searxng`
- run a temp redis instance `docker run   -p 6379:6379 redis`
- bun dev

5. Perform the following tests:
- Create a new search using the SearchXNG backend
- Verify that search results are returned correctly
- Check that image results are included when available
- Test with `SEARXNG_IMAGE_PROXY` set to both true and false
6. Set `SEARCH_API` to other options (tavily, exa) and verify that the original search functionality still works.

## Potential Impacts

- No breaking changes to existing functionality.
- Added flexibility in search engine choice.
- Slight changes to search result processing to accommodate SearchXNG's response format.

## Additional Notes

- This PR modifies the core search logic but maintains compatibility with existing search providers.
- Performance impact should be minimal, with potential improvements for users running local SearchXNG instances.
- Future considerations may include adding more SearchXNG-specific features and optimizing the integration.

## Checklist

- [x] Updated README.md with SearchXNG setup instructions
- [x] Updated .env.local.example with new environment variables
- [x] Performed manual testing with SearchXNG and other search providers
- [x] Reviewed and updated code comments
- [x] Updated docker-compose.yml to include SearchXNG service
- [x] included searxng-settings.yml to help configure searxng docker settings.
- [x] included optional searxng-limiter.toml


Closes #[196] Provide local Search API option
